ENH: Add the ability to have a separate title for each subplot when plotting

icanhazcodeplz-zz opened the pull request:
- [] tests added / passed
 - [] passes ``git diff upstream/master | flake8 --diff``
 - [] whatsnew entry

icanhazcodeplz-zz committed:
Add logic such that if 'title' is a list and 'subplots' is True, use each item of the list as the title of the individual subplots.

icanhazcodeplz-zz committed:
ENH: Add the ability to have a separate title for each subplot when plotting.

icanhazcodeplz-zz committed:
Merge branch 'master' into plotting

sinhrks commented:
Thx for the PR. can u add tests and release note? Also, pls link the original issue if we have.

icanhazcodeplz-zz committed:
switch to using 'is_list_like'

icanhazcodeplz-zz committed:
add line describing the new titles for subplots feature

icanhazcodeplz-zz committed:
add tests to check if the titles above subplots are correct if 'subplots' == True and 'title' is a list

bmagnusson commented on pandas/tools/plotting.py:
```diff
@@ -1217,7 +1217,11 @@ def _adorn_subplots(self):
         if self.title:
             if self.subplots:
-                self.fig.suptitle(self.title)
+                if type(self.title) == list:
```
Ok I'll use `is_list_like`. I purposely did not check the lengths as a feature. As it's written, if you only provide a list of two strings, but the plot has 3 suplots, the first two subplots will have a title and the third will be left without one.

bmagnusson commented:
I did my best at adding a test and release note. I didn't see an original issue that this addresses. It's just something I have on my local that I wanted to contribute.

sinhrks commented on pandas/tests/plotting/test_misc.py:
```diff
@@ -277,6 +274,24 @@ def test_radviz(self):
+        self.assertEqual([p.title._text for p in plot], title)
+
+        # Case len(title) > len(df)
+        plot = df.plot(subplots=True, title=title + ['Ignore me!'])
```
I think we should raise if length of passed list and number of axes are different.

bmagnusson replied:
Sounds good I added exceptions and pushed. Is ValueError appropriate here?

sinhrks commented on pandas/tools/plotting.py:
```diff
@@ -1217,7 +1217,11 @@ def _adorn_subplots(self):
+                else:
+                    self.fig.suptitle(self.title)
             else:
                 self.axes[0].set_title(self.title)
```
should raise if ``subplots=False`` and input is list-like

icanhazcodeplz-zz committed:
-raise ValueError if len(title) != number of columns.
-raise ValueError if subplots=False and title is of type list

bmagnusson commented:
Is there anything else I need to do for this change to get pulled in?

sinhrks commented on pandas/tools/plotting.py:
```diff
@@ -1217,8 +1217,22 @@ def _adorn_subplots(self):
             if self.subplots:
-                self.fig.suptitle(self.title)
+                if is_list_like(self.title):
+                    if len(self.title) != len(self.axes):
```
pls use  ``self.nseries``. Number of columns and axes and can differ when ``layout`` is specified.

Also add test case with 3 numeric columns df with ``df.plot(subplots=True, layout=(2, 2))``

bmagnusson replied:
Done :)

sinhrks commented:
@TomAugspurger pls have a look when you have a time.

TomAugspurger commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -40,7 +40,7 @@ Other enhancements
 
 - ``pd.read_excel`` now preserves sheet order when using ``sheetname=None`` (:issue:`9930`)
-
+- ``pd.DataFrame.plot`` now prints a title above each subplot if ``suplots=True`` and ``title`` is a list of strings
```
You can add this PR number in the issue tag (:issue:`14753`)

bmagnusson replied:
Done :)

TomAugspurger commented on pandas/tools/plotting.py:
```diff
@@ -1217,8 +1217,22 @@ def _adorn_subplots(self):
-                self.fig.suptitle(self.title)
+                if is_list_like(self.title):
+                    if len(self.title) != len(self.axes):
+                        msg = 'The length of `title` must equal the number ' \
```
It'd be nice to report the length of `title` and the expected number of columns in the error message.

bmagnusson replied:
Done :)

icanhazcodeplz-zz committed:
added :issue:`14753`

icanhazcodeplz-zz committed:
-use self.nseries instead of len(self.axes)
-print out length of title and number of columns in error message

icanhazcodeplz-zz committed:
-use self.nseries instead of len(self.axes)
-print out length of title and number of columns in error message
-added test case for when layout=(2,2) but number of columns=3

icanhazcodeplz-zz committed:
Merge remote-tracking branch 'upstream/master' into plotting

# Conflicts:
#	doc/source/whatsnew/v0.20.0.txt

TomAugspurger commented on pandas/tests/plotting/test_misc.py:
```diff
@@ -277,6 +274,32 @@ def test_radviz(self):
+
+        # Case len(title) == len(df)
+        plot = df.plot(subplots=True, title=title)
+        self.assertEqual([p.title._text for p in plot], title)
```
Sorry, I missed this earlier, but can you use `p.get_title()` instead of the private matplotlib `._text` method. Same thing down on line 300, `ax.get_title()`.

bmagnusson replied:
Oh cool I didn't know that existed! Just pushed with the change.

icanhazcodeplz-zz committed:
use .get_title() instead of .title._text

TomAugspurger commented:
Great thanks. +1 for merge when travis passes. Just ping us if you notice that it's green before we do

bmagnusson commented:
All green :)

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -1217,8 +1217,25 @@ def _adorn_subplots(self):
+                        msg = 'The length of `title` must equal the number ' \
+                              'of columns if using `title` of type `list` ' \
+                              'and `subplots=True`.\n' \
+                              'length of title = {}\n' \
```
Can you use parentheses around the full string instead of `\` for the line continuation? (for consistency within the project)

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -1217,8 +1217,25 @@ def _adorn_subplots(self):
             else:
+                if is_list_like(self.title):
+                    msg = 'Using `title` of type `list` is not supported ' \
+                          'unless `subplots=True` is passed'
```
same here

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -2555,8 +2572,10 @@ def _plot(data, x=None, y=None, subplots=False,
-        Title to use for the plot
+    title : string or list
+        If a string is passed, print the string at the top of the figure. If a
+        list is passed and subplots is True, print each item in the
```
single backticks (`` ` ``) around 'subplots' (to make it clear that it is a keyword)

jorisvandenbossche commented on pandas/tools/plotting.py:
```diff
@@ -2555,8 +2572,10 @@ def _plot(data, x=None, y=None, subplots=False,
-    title : string
-        Title to use for the plot
+    title : string or list
+        If a string is passed, print the string at the top of the figure. If a
```
Can you leave the starting sentence "Title to use for the plot" ?

jorisvandenbossche reviewed (commented):
@bmagnusson few more minor (stylistic) comments

icanhazcodeplz-zz committed:
-Add 'Title to use for the plot' as the first sentence for the docstring.
-A few other stylistic changes for consistency.

bmagnusson commented:
Huh. I assume there is something funny going on since I didn't change any actual code. Is there a way to restart the checks?

icanhazcodeplz-zz committed:
Merge remote-tracking branch 'upstream/master' into plotting

icanhazcodeplz-zz committed:
Fix 'continuation line under-indented for visual indent' found by linting.

bmagnusson commented:
Ah! I figured it out. I had a linting error (I'll start doing lint before committing oops!)

bmagnusson commented:
All green now! Anything else to change? If not, thanks for all the help everyone! I had fun. This was my first pull request but it certainly won't be the last :).

jorisvandenbossche commented:
@bmagnusson Thanks! Looking forward to more PRs :-)
